### ISSUE TITLE
aarch64-unknown-none-softfloat: ABI unsoundness when enabling "neon" feature

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/rust#134375]
Enabling the "neon" target feature on the aarch64-unknown-none-softfloat target is taken by LLVM as a sign that we want to use the hardfloat ABI. That is unfortunate as it makes it UB to link such code against code built for aarch64-unknown-none-softfloat without the "neon" target feature. (Note that it's not "neon" which is problematic but "fp-armv8"; however, the two are tied together by rustc.)

For Rust-generated functions we work around this by forcing our own ABI, passing floats either indirectly or via integer registers (https://github.com/rust-lang/rust/pull/133102). However, this does not help for LLVM-generated calls for builtins/intrinsics, as shown in [this example](https://godbolt.org/z/r56xWo8nT) by @beetrees.

We don't have target maintainers listed for this target, so maybe that means we can just demote it to tier 3? (See https://github.com/rust-lang/rust/issues/113739)

LLVM issue: https://github.com/llvm/llvm-project/issues/110632. So far LLVM maintainers seem to not agree that there is a problem here.

@Amanieu unfortunately your proposal for making floats work on that target doesn't quite suffice. :/ We do need some help from LLVM.
@nikic do you have any good ideas for what we could do here? It seems like rejecting enabling "neon" on aarch64-unknown-none-softfloat is the only sound option we have right now, but I worry that may make the Rust-for-Linux folks (among others) unhappy.

**Comment by apiraino** [comment]
> We don't have target maintainers listed for this target, so maybe that means we can just demote it to tier 3?

I also skimmed the git history for the https://github.com/rust-lang/rust/blob/master/src/doc/rustc/src/platform-support.md but found nothing. I think it could make sense to demote the target if it's unmaintained. @RalfJung should I mention this question in the next compiler meeting?

Another question: do you think the `neon` target feature could create undefined behaviours also in other targets? (I don't know how they work)

thanks

**Comment by RalfJung** [comment]
<https://github.com/rust-lang/rust/issues/131058> was about various issues with the neon target feature; except for this one, they should all be resolved.

**Comment by RalfJung** [comment]
With https://github.com/rust-lang/rust/pull/134794 we could fairly easily reject enabling `neon` specifically on this target. That would fix the ABI issues.

I just wonder, is anyone relying on enabling `neon` for `aarch64-unknown-none-softfloat`? Not quite sure whom to ping... @Darksonn for R4L, who else might be using this target?

**Comment by ojeda** [comment]
Cc @JamieCunliffe (currently, arm64 in the kernel uses `--target=aarch64-unknown-none -Ctarget-feature="-neon"`).

**Comment by Darksonn** [comment]
Should we be using `--target=aarch64-unknown-none-softfloat` instead?

**Comment by RalfJung** [comment]
If you never need `neon`, that would be slightly better... but there's still no sound way to enable `neon` (or `fp-armv8`) for some of the code. The limiting factor here is LLVM (https://github.com/llvm/llvm-project/issues/110632). The C ecosystem assumes that there is a hard-error when floats are used somewhere that `neon` is not available, so there's no proper support for a softloat ABI. However Rust assumes float types always exist and there's some reasonable ABI for them.

Probably, someone will have to add a `soft-float` target feature to LLVM that forces the use of integer registers for passing floats even when `fp-armv8` is available.

**Comment by RalfJung** [comment]
@Darksonn @ojeda does Rust4Linux need to locally enable `neon` for some parts of the build? Or is that all going to stay on the C side for the foreseeable future?

Right now the only way I can see to fix this issue is to just mark the `neon` target feature as incompatible with `aarch64-unknown-none-softfloat`. Some work on the LLVM side is needed to fix that in a way that is compatible with Rust (where we want to always provide the float types and just use softfloats when needed). That said, the C approach of just using different compiler flags for different parts of the code and "being careful" when linking (making sure there are no float types crossing that ABI boundary) would also work in Rust.

**Comment by asahilina** [comment]
I only use floats in const context (to transmute f32 to u32) and don't enable neon or anything like that. I didn't even realize Rust had an aarch64 softfloat target, I ended up implementing my own softfloat in pure integer Rust instead where I needed float math.

**Comment by RalfJung** [comment]
https://github.com/rust-lang/rust/pull/135160 proposes to make `#[target_feature(enable = "neon")]` a hard error on `aarch64-unknown-none-softfloat`; please chime in if that would be a problem for anyone.

**Comment by briansmith** [comment]
> I don't think the existing C users will be rewritten anytime soon, but who knows what out-of-tree may be doing.

It is inevitable that *some* (perhaps third party / out of tree) Rust code will need to use NEON within the Linux kernel, in other kernels, and in similar situations. It is just a matter of time.

In the same way that people are replacing C code with Rust code, people are also trying to replace out-of-line assembly with smaller bits of inline assembly and/or with auto-vectorized code, using patterns like this:

```
fn foo(safe_to_use_vectorized: bool) {
    // TODO(MSRV): "neon" isn't supported by Rust for target_arch = "arm" yet.
    #[cfg(any(target_arch = "x86", target_arch = "x86_64", target_arch = "aarch64"))]
    if safe_to_use_vectorized {
        #[cfg_attr(any(target_arch = "aarch64"), target_feature(enable="neon"))]
        #[cfg_attr(any(target_arch = "x86", target_arch = "x86_64"), target_feature(enable="sse2"))]
         unsafe fn foo_vectorized() { foo_inner() } // auto-vectorized foo_inner    
     
        return unsafe { foo_vectorized() };
    }
    
    let _ = safe_to_use_vectorized;
    foo_inner()
}

#[inline(always)] // So it can be auto-vectorized within `foo_vectorized()`
fn foo_inner() {
    // auto-vectorizable code goes here.
}
```

For people writing code like the above, we'd need to change our `target_arch = "aarch64"` conditions to something so that it can build for `aarch64-unknown-none-softfloat`. I.e. What is the `cfg()` expression for 'an aarch64 target that supports `target_feature(enable="neon")`'?

So, I don't think that disabling `target_feature(enable="neon")` for these targets isn't a long-terms solution to the ABI issues.

**Comment by RalfJung** [comment]
>  What is the cfg() expression for 'an aarch64 target that supports target_feature(enable="neon")'?

`all(target_arch="aarch64", not(target_abi="softfloat"))`

> So, I don't think that disabling target_feature(enable="neon") for these targets isn't a long-terms solution to the ABI issues.

Did you intend the double negation? This sentence says you do think that it is a long-term solution...

To be clear, I don't think this is a long-term solution. But it also makes no sense to pretend to support something which our primary backend does not support. Someone has to put in the work on the LLVM side before we can provide this feature to our users.

**Comment by briansmith** [comment]
> Did you intend the double negation? This sentence says you do think that it is a long-term solution...

Obviously not. I edited it.

> To be clear, I don't think this is a long-term solution. But it also makes no sense to pretend to support something

We agree on that.

**Comment by briansmith** [comment]
> However, this does not help for LLVM-generated calls for builtins/intrinsics, as shown in [this example](https://godbolt.org/z/r56xWo8nT) by @beetrees.

The example is, in part:
```
#[inline(never)]
#[target_feature(enable = "neon")]
pub unsafe fn yes_neon(x: u128) -> f64 {
    x as f64
}```

Is there an example that doesn't use floating point types within a `#[target_feature(enable = "neon")]` function?

If not, it would be better to "just" add a check (maybe even just a lint) that rejects any use of floating point types within `#[target_feature(enable = "neon")]` functions, when the target doesn't enable NEON by default, instead of rejecting [target_feature(enable = "neon")]` completely. This would then allow everything to work, with probably no disruption.

**Comment by RalfJung** [comment]
I am not confident enough in our ability to judge when LLVM will insert function calls whose ABI might be affected, to be comfortable with that approach.

**Comment by briansmith** [comment]
> I am not confident enough in our ability to judge when LLVM will insert function calls whose ABI might be affected, to be comfortable with that approach.

I think we would improve our comfort there.

It seems like the issue with intrinsics is that either LLVM is making assumptions about intrinsics that aren't valid, and/or Rust is using intrinsics in a way that isn't compatible with how they are designed in LLVM.

It seems like rustc is already able to handle cross-ABI function calls--either it has machinery for an enable-NEON function can call a non-enable-NEON function, or it has machinery to reject those calls--but that machinery isn't being invoked for calls to LLVM intrinsics. Or maybe the machinery is being invoked if the call to the intrinsics is done in actual Rust code but not if the call to the intrinsic is generated by rustc? If that is the case, then that seems like something to be improved, and naively it seems like it would be practical to make that improvement.

**Comment by briansmith** [comment]
In particular, in the actual (C) runtime library, there is object code for each called function. That means there is a function prototype for that function within LLVM/libc that describes its interface. But that declaration apparently doesn't have an adequate description of the interface, and/or internally LLVM doesn't have adequate checking of the ABI compatibility when inserting the calls.

Also, there are very few (non-inlined) functions in the runtime library (object code) that take SIMD and/or floating point types as arguments and/or return them. And very few places that insert calls to those few functions. So I do think it is quite a tractable problem to solve even through brute force.

**Comment by briansmith** [comment]
> If not, it would be better to "just" add a check (maybe even just a lint) that rejects any use of floating point types within `#[target_feature(enable = "neon")]` functions, when the target doesn't enable NEON by default, instead of rejecting `#[target_feature(enable = "neon")]` completely. This would then allow everything to work, with probably no disruption.

This seems to be exactly what clang does for C code when using `-march=armv8+nofp`; see [Godbolt](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:___c,selection:(endColumn:1,endLineNumber:17,positionColumn:1,positionLineNumber:17,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:'__attribute__((noinline))+%0A__attribute__((target(%22neon%22)))%0A__uint128_t+neon_no_float(__uint128_t+x)+%7B%0A++++return+x%3B%0A%7D%0A%0A__attribute__((noinline))+%0Afloat+no_neon_float(__uint128_t+x)+%7B%0A++++return+(float)x%3B%0A%7D%0A%0A__attribute__((noinline))+%0A__attribute__((target(%22neon%22)))%0Afloat+neon_float(__uint128_t+x)+%7B%0A++++return+(float)x%3B%0A%7D%0A'),l:'5',n:'0',o:'C+source+%231',t:'0')),k:34.489528795811516,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((g:!((h:compiler,i:(compiler:armv8-cclang1910,filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1',verboseDemangling:'0'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:___c,libs:!(),options:'-march%3Darmv8%2Bnofp',overrides:!(),selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+armv8-a+clang+19.1.0+(Editor+%231)',t:'0')),k:50,l:'4',m:50,n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(compilerName:'rustc+nightly',editorid:1,fontScale:14,fontUsePx:'0',j:1,wrap:'1'),l:'5',n:'0',o:'Output+of+armv8-a+clang+19.1.0+(Compiler+%231)',t:'0')),header:(),k:50,l:'4',n:'0',o:'',s:0,t:'0')),k:65.51047120418848,l:'2',m:100,n:'0',o:'',t:'0')),l:'2',n:'0',o:'',t:'0')),version:4).
```
__attribute__((noinline)) 
__attribute__((target("neon")))
__uint128_t neon_no_float(__uint128_t x) {
    return x;
}

__attribute__((noinline)) 
float no_neon_float(__uint128_t x) {
    return (float)x;
}

__attribute__((noinline)) 
__attribute__((target("neon")))
float neon_float(__uint128_t x) {
    return (float)x;
}
```
It accepts "neon_no_float" even though it enables NEON, because it doesn't use floating point types. It rejects "no_neon_float" even though it doesn't use NEON, because it uses `float`. It rejects `neon_float` because it uses `float`.

**Comment by RalfJung** [comment]
The issue is that LLVM does not have a concept of different ABIs for aarch64. It does some ad hoc fallback when certain target features are missing but there is nothing systematic.

This is in contrast to arm-32 and riscv where LLVM has explicit ABI knobs that let us request softfloat abi, and that work reliably no matter the target features. x86 at least has a target feature that lets us force the soft-float ABI no matter what else is available; that is far from great but it's something. Aarch64 has nothing.

This cannot be fixed without some LLVM work.

**Comment by RalfJung** [comment]
The issue is that LLVM does not have a concept of different ABIs for aarch64. It does some ad hoc fallback when certain target features are missing but there is nothing systematic.

This is in contrast to arm-32 and riscv where LLVM has explicit ABI knobs that let us request softfloat abi, and that work reliably no matter the target features. x86 at least has a target feature that lets us force the soft-float ABI no matter what else is available; that is far from great but it's something. Aarch64 has nothing.

This cannot be fixed without some LLVM work.

> If not, it would be better to "just" add a check (maybe even just a lint) that rejects any use of floating point types within #[target_feature(enable = "neon")] functions, when the target doesn't enable NEON by default, instead of rejecting #[target_feature(enable = "neon")] completely. This would then allow everything to work, with probably no disruption.

Aside from this being fragile re: incomplete knowledge about the ways in which LLVM likes to screw us over, Rust also has historically rejected the idea of making floats or float-ops only available in a target-dependent way.

**Comment by briansmith** [comment]
> Aarch64 has nothing. This cannot be fixed without some LLVM work.

But why should LLVM ever add support for it? Basically, we're asking LLVM to add support for calling these intrinsic-backing functions for floating point operations that nobody actually wants to use, so basically add a bunch of dead code to LLVM, that will used primarily by the Rust toolchain test suite and nothing else. That doesn't make sense. In other words, I don't think there's anything wrong with LLVM's approach, so I don't see why they would change it.

Is there actually any useful code that uses floating point in an aarch64-unknown-none target, at all? Is there any existing floating point code in the Linux kernel, at all that would hint at the need for floating point support in Rust for the Linux kernel or any OS kernel? AFAICT, no.

Conversely, there's tons of NEON code in the Linux kernel and other ARM/ARM64 kernels, currently written in out-of-line assembly, unfortunately.

> Rust also has historically rejected the idea of making floats or float-ops only available in a target-dependent way.

Sure, but that's a rustc problem, not an LLVM problem. And even if LLVM fixed it, it wouldn't be fixed for non-LLVM backends. Instead of changing every backend, I think it makes more sense to change the interface between rustc and the backend so that rustc can divert its float support to a library for these targets. Even if this hurts the optimization of the floating point operations when the diversion is done, it will be OK, because ~nobody is actually going to be using it.

**Comment by briansmith** [comment]
> Conversely, there's tons of NEON code in the Linux kernel and other ARM/ARM64 kernels, currently written in out-of-line assembly, unfortunately.

To clarify, there is a lot of explicit NEON-using code written in C in the Linux kernel, as well.

**Comment by RalfJung** [comment]
If LLVM doesn't intend to support aarch64-softfloat, we should remove the target entirely and instead add aarch64-nofloat -- but that will require an RFC that explains how we handle targets without float support.

**Comment by nikic** [comment]
FWIW, I started looking into this and doing the obvious thing mostly works, but I ran into issues with legalization of f128 libcalls. The problem is basically that for AArch64 f128 is a legal type, but i128 is not, so libcall legalization would have to directly go from a pair of i64 to f128 via stack load+store I think. Dealing with that kind of issue will be very annoying :(

**Comment by nikic** [comment]
The `-fp-armv8` case is easy because `f128` is not a legal type, so everything can stay in integers.

**Comment by RalfJung** [comment]
So even `fadd` or `fptoui` just does not work on `f128` with `-fp-armv8`? It doesn't call some softfloat routines?

**Comment by tgross35** [comment]
> for AArch64 f128 is a legal type, but i128 is not

What makes a type legal - is it ABI specification, or the ability to be passed as a single register?

**Comment by RalfJung** [comment]
> which GCC/clang target ABI is link-/runtime-compatible with Rust's aarch64-softfloat ABI?

I don't know (I would recommend not passing anything over the ABI boundary that would usually use one of the missing registers), but interaction with C code is a separate issue from this one which is about things already being unsound entirely on the Rust side.

**Comment by briansmith** [comment]
Sure, but my point is that asking LLVM to create a softfloat target for this will only help if LLVM is also guaranteeing that that softfloat target is ABI-compatible with the nofloat target, what people need is Rust compatibility with the nofloat target. And solutions to this issue that aren't compatible with the nofloat target are a non-starter, right?

**Comment by RalfJung** [comment]
I would strongly assume that the proper softfloat ABI (if it ever exists) will be identical to what LLVM already does when all the FPU and SIMD feature gates are turned off. But this is really quite speculative as no such ABI exists yet, so... not quite sure what the point is of discussing this at this time.

**Comment by RalfJung** [comment]
Cc @davidtwco @JamieCunliffe for some ARM folks -- if you have contacts with the LLVM aarch64 team, would be great to get this fixed properly on the LLVM side. :D For now we'll be nerfing the aarch64 softfloat target in Rust to avoid issues here (see https://github.com/rust-lang/rust/pull/135160), but that's of course not a proper solution.

**Comment by RalfJung** [comment]
This is still the tracking issue for the lint.

**Comment by pitust** [comment]
This future incompatibility warning triggers on libcore, in `/home/pitust/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/../../stdarch/crates/core_arch/src/aarch64/neon/generated.rs`, when compiling for this target spec:
```json
{
    "abi": "softfloat",
    "arch": "aarch64",
    "code-model": "small",
    "cpu": "cortex-a57",
    "crt-objects-fallback": "false",
    "data-layout": "e-m:e-p270:32:32-p271:32:32-p272:64:64-i8:8:32-i16:16:32-i64:64-i128:128-n32:64-S128-Fn32",
    "eh-frame-header": true,
    "emit-debug-gdb-scripts": true,
    "features": "+v8a,+strict-align,-neon,-fp-armv8",
    "linker": "rust-lld",
    "linker-flavor": "gnu-lld",
    "llvm-abiname": "softfloat",
    "llvm-target": "aarch64-elf",
    "max-atomic-width": 64,
    "panic-strategy": "abort",
    "supported-sanitizers": [],
    "target-pointer-width": "64",
    "position-independent-executables": false,
    "dynamic-linking": false,
    "target-mcount": "__stack_probe",
    "dll-prefix": "",
    "dll-suffix": ""
}```

I think this is mostly about the fact that I am using the soft float ABI, but I'm not sure.

**Comment by apiraino** [comment]
Assigning P-high (discussion on [Zulip](https://rust-lang.zulipchat.com/#narrow/channel/245100-t-compiler.2Fprioritization.2Falerts/topic/.23134375.20aarch64-unknown-none-softfloat.3A.20ABI.20unsoundness.20whe.E2.80.A6/with/519972563)) until we figure out what to do (fix, demote or something else)

@rustbot label -I-prioritize +P-high

**Comment by newpavlov** [comment]
Relevant issue: #117938

I think the compiler should implement this [suggestion](https://github.com/rust-lang/rust/issues/117938#issuecomment-1813236792):

>Maybe we should declare (and have the feature-detect macros implement) that SSE features are never available on softfloat targets. Then we can compile functions with SSE #[target_features] into unreachable_unchecked and so their ABI does not matter so we can generate whatever LLVM IR we want.

Instead of making it a hard error, which would be a breaking change on softfloat ARM targets for a number of widely used crates (`aes`, `sha2`, `keccak`, etc.).

**Comment by RalfJung** [comment]
Interesting... I never made a connection between that issue and this. This issue is an ABI problem, resolved by rustc better understanding the requirements for proper ABI modeling in LLVM to avoid emitting nonsensical target feature combinations, which is the subsystem that triggers the warning you saw. That other issue is LLVM crashing on combinations of target features it doesn't like even if those target features actually make perfect sense.

But you are right that the system we use to work around that LLVM bug may actually also be useful here. I'll make a new issue for that since it's a wider discussion, not specific to any particular architecture / ABI.

**Comment by RalfJung** [comment]
Let's use https://github.com/rust-lang/rust/issues/160301 to discuss this alternative way of dealing with ABI-incompatible target features in `#[target_feature]`.
